kvserver: don't let tenants retrieve KV-level trace messages

tbg opened the pull request:
This strips `BatchResponse` of the verbose logs when returned to a tenant.

Touches https://github.com/cockroachdb/cockroach/issues/58610.

Release note: None

tbg committed:
kvtenantccl: test that tenants can't see KV-level trace messages

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70406)
<!-- Reviewable:end -->

tbg commented:
@nvanbenschoten this isn't by design, right? I thought we'd strip the traces at the KV->Tenant boundary. I don't think we can let the tenants have unredacted host cluster traces, with the exception of `ContentionEvent`.

tbg committed:
kvserver: strip verbose logs and tags for tenant traces

Touches #70407.

It does not exactly fix it, as the approach here is a blocklist on our
main KV endpoint, whereas we really need an allowlist that applies to
all possible current and future endpoints.

Release note: None

nvb reviewed (approved):
This <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, though I share the same concerns as @irfansharif regarding the impact that this will have on the observability of KV's handling of requests in serverless clusters. There is a large class of support issues that will become significantly harder to diagnose without these traces, like those related to contention or those related to MVCC garbage. This does seem like a reasonable first step, but we'll want to continue to push on this, and soon.

I think @andreimatei should also sign off on this, given his focus on tracing.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70406#-:-Mk2yjxH6MM8pZFr_Gw0:bg9j8dq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/server/node.go, line 1049 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70406#-Mk2yExw30dUrQs6ZZZQ:-Mk2yExw30dUrQs6ZZZR:bb6huk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa5b88472c38e7ba1feeea8c0bedd2d20b8f686e/pkg/server/node.go#L1049)):*
> ```Go
> 		br.Error = roachpb.NewError(err)
> 	}
> 	if tenantID != roachpb.SystemTenantID {
> ```

Should this live in the closure returned from `setupSpanForIncomingRPC`?


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
LGTM too, and thanks for backfilling your comment on #58610 around possible paths forward to bring back usable KV trace data for secondary tenants. I'm curious if there are other ideas; it'd be nice to address soon.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70406#-:-Mk3-_-z-MbyP6ET-UC9:b-pq87oe)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70406)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/server/node.go, line 1049 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70406#-Mk2yExw30dUrQs6ZZZQ:-Mk311xQCzeM3IfjcBeh:b47avhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa5b88472c38e7ba1feeea8c0bedd2d20b8f686e/pkg/server/node.go#L1049)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this live in the closure returned from `setupSpanForIncomingRPC`?
</blockquote></details>

+1, because also streaming and other RPCs
___
*[pkg/server/node.go, line 1054 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70406#-Mk30pfL58lEzefDifBy:-Mk30pfL58lEzefDifBz:b7j4eq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa5b88472c38e7ba1feeea8c0bedd2d20b8f686e/pkg/server/node.go#L1054)):*
> ```Go
> 		for i := range br.CollectedSpans {
> 			// Guard against a new sensitive field being added to RecordedSpan.
> 			type calcifiedRecordedSpan struct {
> ```

let's do this static check in a test instead
___
*[pkg/server/node.go, line 1070 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70406#-Mk31hg07RWyAwWnbsA5:-Mk31hg07RWyAwWnbsA6:b7bs02a) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa5b88472c38e7ba1feeea8c0bedd2d20b8f686e/pkg/server/node.go#L1070)):*
> ```Go
> 			}
> 			sp := (*calcifiedRecordedSpan)(&br.CollectedSpans[i])
> 			sp.Tags = nil
> ```

Was it intentional that you're leaving the span structure in place, instead of taking all the `StructuredRecords` and moving them to a single span (or even to a new field on the `BatchResponse`? I can't decide if the spans without messages are worth having.

And btw, are the keys in `ContentionEvets` safe? Can a `ContentionEvent` come from a meta range? I don't see why not exactly. Depending on the security model of the SQL pods, maybe returning these such events is not a problem since a user doesn't get to see them (?), but then again maybe we should strip out all events with keys not belonging to the tenant.

I've been thinking (and more so after talking to Irfan too) that perhaps the way in which we've built this event collection on top of tracing is not great because currently collecting contention events for all queries (which I think we should do) would be too expensive because it involves creating a structured trace. These events don't need that structure; all the events for a RPC/query could go to a shared container rather than going to different spans. So this might be a hint that introrucing an events field on the `BatchResponse` (or perhaps in some out-of-band gRPC metadata) might be a good thing to do to give us flexibility in the future.
___
*[pkg/server/node.go, line 1071 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70406#-Mk31s9S9X-KnD4YJBwA:-Mk31s9S9X-KnD4YJBwB:byi9hn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa5b88472c38e7ba1feeea8c0bedd2d20b8f686e/pkg/server/node.go#L1071)):*
> ```Go
> 			sp := (*calcifiedRecordedSpan)(&br.CollectedSpans[i])
> 			sp.Tags = nil
> 			sp.Logs = nil
> ```

let's put a `redacted` message in `sp.Logs`


<!-- Sent from Reviewable.io -->

tbg commented:
Superseded by https://github.com/cockroachdb/cockroach/pull/70562
